pkg/http/sender.go

```diff
@@ -0,0 +1,638 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2022 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package http
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"math/big"
+	"math/rand"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/questdb/go-questdb-client/v3/pkg/buffer"
+	"github.com/questdb/go-questdb-client/v3/pkg/conf"
+)
+
+type tlsMode int64
+
+const (
+	tlsDisabled           tlsMode = 0
+	tlsEnabled            tlsMode = 1
+	tlsInsecureSkipVerify tlsMode = 2
+)
+
+var (
+	// We use a shared http transport to pool connections
+	// across LineSenders
+	globalTransport http.Transport = http.Transport{
+		Proxy:                 http.ProxyFromEnvironment,
+		MaxConnsPerHost:       0,
+		MaxIdleConns:          100,
+		MaxIdleConnsPerHost:   100,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+		TLSClientConfig:       &tlsClientConfig,
+	}
+
+	// clientCt is used to track the number of open LineSenders
+	// If the clientCt reaches 0, meaning all LineSenders have been
+	// closed, the globalTransport closes all idle connections to
+	// free up resources
+	clientCt atomic.Int64
+
+	// Since we use a global http.Transport, we also use a global
+	// tls config. By default, this is set to verify all server certs,
+	// but can be changed by the WithTlsInsecureVerify setting.
+	tlsClientConfig tls.Config = tls.Config{}
+)
+
+// LineSender allows you to insert rows into QuestDB by sending ILP
+// messages over HTTP(S).
+//
+// Each sender corresponds to a single HTTP client. All senders
+// utilize a global transport for connection pooling. A sender
+// should not be called concurrently by multiple goroutines.
+type LineSender struct {
+	buffer.Buffer
+
+	address string
+
+	// Retry/timeout-related fields
+	retryTimeout                time.Duration
+	minThroughputBytesPerSecond int
+	requestTimeout              time.Duration
+
+	// Auto-flush fields
+	autoFlush     bool
```

Comment:
We don't need this flag in the sender. `autoFlushRows == 0` could mean that auto flush is disabled.